openSearchLane.go

```diff
@@ -15,50 +16,78 @@ import (
 	"github.com/opensearch-project/opensearch-go/v3/opensearchapi"
 )
 
-type LogTemplate struct {
+const (
+	DefaultLogThreshold    = 5
+	DefaultMaxBufferSize   = 10
+	DefaultBackoffInterval = 10 * time.Second
+	DefaultBackOffLimit    = 10 * time.Minute
+)
+
+type OpenSearchLogMessage struct {
 	AppName      string            `json:"appName"`
 	ParentLaneId string            `json:"parentLaneId,omitempty"`
 	JourneyID    string            `json:"journeyId,omitempty"`
-	LaneID       string            `json:"laneId"`
-	LogMessage   string            `json:"logMessage"`
+	LaneID       string            `json:"laneId,omitempty"`
+	LogMessage   string            `json:"logMessage,omitempty"`
 	Metadata     map[string]string `json:"metadata,omitempty"`
 }
 
+type EmergencyFn func(logBuffer any, logPath string)
+
+type OpenSearchLaneConfig struct {
+	LogThreshold    int           `json:"logThreshold,omitempty"`
+	MaxBufferSize   int           `json:"maxBufferSize,omitempty"`
+	BackoffInterval time.Duration `json:"backoffInterval,omitempty"`
+	BackOffLimit    time.Duration `json:"backoffLimit,omitempty"`
+	LogPath         string        `json:"logPath,omitempty"`
+}
+
 type openSearchLane struct {
 	logLane
 	openSearchConnection *openSearchConnection
-	wg                   sync.WaitGroup
 	metadata             map[string]string
 }
 
 type openSearchConnection struct {
-	client     *opensearchapi.Client
-	mu         sync.Mutex
-	logBuffer  []*LogTemplate
-	refCountCh chan *sync.WaitGroup
-	refCount   int
-	index      string
-	appName    string
+	client         *opensearchapi.Client
+	mu             sync.Mutex
+	logBuffer      []*OpenSearchLogMessage
+	stopCh         chan *sync.WaitGroup
+	wakeCh         chan struct{}
+	refCount       int
+	index          string
+	appName        string
+	config         OpenSearchLaneConfig
+	emergencyFn    EmergencyFn
+	ctx            context.Context
+	cancelFn       context.CancelFunc
+	messagesQueued int
+	messagesSent   int
 }
 
 type OpenSearchLane interface {
 	Lane
-	Connect(openSearchUrl, openSearchPort, openSearchUser, openSearchPass, openSearchIndex, openSearchAppName string) (err error)
-	SetEmergencyHandler()
+	LaneMetadata
+	Connect(openSearchUrl, openSearchPort, openSearchUser, openSearchPass, openSearchIndex, openSearchAppName string, config OpenSearchLaneConfig) (err error)
+	SetEmergencyHandler(emergencyFn EmergencyFn) (err error)
+	OpenSearchLaneStats()
```

Comment:
Since this is inside OpenSearchLane, the "go" way is brevity, so `Stats()` instead of `OpenSearchLaneStats()`